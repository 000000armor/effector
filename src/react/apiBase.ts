import {
  Store,
  is,
  createNode,
  step,
  clearNode,
  scopeBind,
  Scope,
  Unit,
  Event,
  Node,
} from 'effector'
import React from 'react'
import {useSyncExternalStore} from 'use-sync-external-store/shim'
import {useSyncExternalStoreWithSelector} from 'use-sync-external-store/shim/with-selector'
import {throwError} from './throw'
import {createWatch} from './createWatch'
import {withDisplayName} from './withDisplayName'

const stateReader = <T>(store: Store<T>, scope?: Scope) =>
  scope ? scope.getState(store) : store.getState()
const basicUpdateFilter = <T>(upd: T, oldValue: T) => upd !== oldValue
const keysEqual = (a?: readonly any[], b?: readonly any[]) => {
  if (!a || !b || a.length !== b.length) return false

  let isEqual = true

  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) {
      isEqual = false
      break
    }
  }

  return isEqual
}

export function useStoreBase<State>(store: Store<State>, scope?: Scope) {
  if (!is.store(store)) throwError('expect useStore argument to be a store')

  const subscribe = React.useCallback(
    (cb: () => void) => createWatch(store, cb, scope),
    [store, scope],
  )
  const read = React.useCallback(
    () => stateReader(store, scope),
    [store, scope],
  )
  const currentValue = useSyncExternalStore(subscribe, read, read)

  return currentValue
}

export function useUnitBase<Shape extends {[key: string]: Unit<any>}>(
  shape: Shape,
  scope?: Scope,
) {
  const isShape = !is.unit(shape) && typeof shape === 'object'
  const normShape = isShape ? shape : {unit: shape}
  const isList = Array.isArray(normShape)
  const entries = Object.entries(normShape)
  const [current, storeMap, stores] = React.useMemo(() => {
    const initial: Record<string, any> = isList ? [] : {}
    const stores: Store<any>[] = []
    const storeMap: Record<string, string> = {}
    for (const [key, value] of entries) {
      if (is.store(value)) {
        if (stores.includes(value)) {
          throwError(`useUnit store at key "${key}" is already exists in shape`)
        }
        stores.push(value)
        /** note that this allows only one occurence of the store */
        storeMap[(value as any).graphite.id] = key
        initial[key] = stateReader(value, scope)
      } else {
        if (!is.unit(value)) throwError('expect useUnit argument to be a unit')
        initial[key] = scope ? scopeBind(value as Event<any>, {scope}) : value
      }
    }
    return [{ref: initial}, storeMap, stores]
  }, [...entries.flat(), scope])
  const subscribe = React.useCallback(
    (cb: () => void) => {
      const seq = [
        step.compute({
          fn(value, _, stack) {
            const storeId = stack.parent!.node.id
            const storeKey = storeMap[storeId]
            current.ref = isList
              ? [...(current.ref as any[])]
              : {...current.ref}
            current.ref[storeKey] = value
          },
        }),
        /**
         * 'effect' priority cannot be batched
         * so we wait other effects to finish and then run batching
         * this is not an essential to work
         * just a fine tuning of a really low priority (ui) update
         *
         * also note that step.run({fn}) is an alias for
         * step.compute({fn, priority: 'effect'})
         * and since 22 version fn might be ommited if not used
         * */
        step.compute({priority: 'effect'}),
        step.compute({priority: 'sampler', batch: true}),
        step.run({fn: () => cb()}),
      ]
      if (scope) {
        const node = createNode({node: seq})
        const scopeLinks: {[_: string]: Node[]} = (scope as any).additionalLinks
        const storeLinks: Node[][] = []
        stores.forEach(store => {
          const id = (store as any).graphite.id
          const links = scopeLinks[id] || []
          scopeLinks[id] = links
          links.push(node)
          storeLinks.push(links)
        })
        return () => {
          storeLinks.forEach(links => {
            const idx = links.indexOf(node)
            if (idx !== -1) links.splice(idx, 1)
          })
          clearNode(node)
        }
      } else {
        const node = createNode({
          node: seq,
          parent: stores,
          family: {owners: stores},
        })
        return () => {
          clearNode(node)
        }
      }
    },
    [current],
  )
  const read = React.useCallback(
    () => (isShape ? current.ref : current.ref.unit),
    [current],
  )
  return useSyncExternalStore(subscribe, read, read)
}

export function useStoreMapBase<State, Result, Keys extends ReadonlyArray<any>>(
  [configOrStore, separateFn]: [
    configOrStore:
      | {
          store: Store<State>
          keys: Keys
          fn(state: State, keys: Keys): Result
          updateFilter?: (update: Result, current: Result) => boolean
        }
      | Store<State>,
    separateFn?: (state: State, keys: Keys) => Result,
  ],
  scope?: Scope,
): Result {
  let fn: (state: State, keys: Keys) => Result
  let updateFilter: (update: Result, current: Result) => boolean =
    basicUpdateFilter
  let store: Store<State>
  let keys: Keys
  if (separateFn) {
    fn = separateFn
    store = configOrStore as Store<State>
    keys = [] as unknown as Keys
  } else {
    fn = (configOrStore as any).fn
    store = (configOrStore as any).store
    keys = (configOrStore as any).keys
    updateFilter = (configOrStore as any).updateFilter || basicUpdateFilter
  }
  if (!is.store(store)) throwError('useStoreMap expects a store')
  if (!Array.isArray(keys)) throwError('useStoreMap expects an array as keys')
  if (typeof fn !== 'function') throwError('useStoreMap expects a function')

  const subscribe = React.useCallback(
    (cb: () => void) => createWatch(store, cb, scope),
    [store, scope],
  )
  const read = React.useCallback(
    () => stateReader(store, scope),
    [store, scope],
  )

  const stateRef = React.useRef<State>()
  const valueRef = React.useRef<Result>()
  const keysRef = React.useRef(keys)

  const value = useSyncExternalStoreWithSelector(
    subscribe,
    read,
    read,
    state => {
      if (stateRef.current !== state || !keysEqual(keysRef.current, keys)) {
        const result = fn(state, keys)

        stateRef.current = state
        keysRef.current = keys

        // skip update, if undefined
        // just like original store or previous implementation
        if (result !== undefined) {
          valueRef.current = result
        }
      }

      return valueRef.current as Result
    },
    (current, update) => !updateFilter(update, current),
  )

  return value
}
export function useListBase<T>(
  list: Store<T[]>,
  renderItem:
    | {
        keys?: any[]
        fn(item: T, index: number): React.ReactNode
        getKey?: (item: T) => string
      }
    | ((item: T, index: number) => React.ReactNode),
  scope?: Scope,
): React.ReactNode {
  let keys = [] as any[]
  let fn
  let getKey: (item: T) => string
  if (typeof renderItem === 'object' && renderItem !== null) {
    if (renderItem.keys) keys = renderItem.keys
    fn = renderItem.fn
    if (renderItem.getKey) getKey = renderItem.getKey
  } else {
    fn = renderItem
  }
  if (!is.store(list)) throwError('expect useList first argument to be a store')
  if (typeof fn !== 'function')
    throwError("expect useList's renderItem to be a function")
  if (!Array.isArray(keys)) throwError("expect useList's keys to be an array")
  const Item = React.useMemo(() => {
    const Item = withDisplayName(
      `${list.shortName || 'Unknown'}.Item`,
      (
        props:
          | {index: number; keys: any[]; keyVal: never; value: never}
          | {index: never; keys: any[]; keyVal: string; value: T},
      ) => {
        const {index, keys, keyVal, value} = props
        const isKeyed = !!fnRef.current[1]
        if (isKeyed) {
          return fnRef.current[0](value, keyVal as any)
        }
        const item = useStoreMapBase(
          [
            {
              store: list,
              keys: [index, ...keys],
              fn: (list, keys) => list[keys[0]],
            },
          ],
          scope,
        )
        return fnRef.current[0](item, index)
      },
    )
    return React.memo(Item)
  }, [list, scope, !!getKey!])
  const fnRef = React.useRef([fn, getKey!] as const)
  fnRef.current = [fn, getKey!]
  const keysSelfMemo = React.useMemo(() => keys, keys)
  if (getKey!) {
    return useStoreBase(list, scope).map(value => {
      const key = fnRef.current[1](value)
      return React.createElement(Item, {
        keyVal: key,
        key,
        keys: keysSelfMemo,
        value,
      })
    })
  } else {
    const length = useStoreMapBase(
      [
        {
          store: list,
          keys: [list],
          fn: list => list.length,
        },
      ],
      scope,
    )
    return Array.from({length}, (_, i) =>
      React.createElement(Item, {
        index: i,
        key: i,
        keys: keysSelfMemo,
      }),
    )
  }
}
