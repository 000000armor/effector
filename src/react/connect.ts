import React from 'react'

import {Store} from 'effector'
import {createReactState} from './createReactState'

export const connect = <State>(Component: React.ComponentType<any>) => (
  store: Store<State>,
) => createReactState(store, Component)
