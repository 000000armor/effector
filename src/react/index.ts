export {createStoreConsumer} from './createStoreConsumer'
export {createComponent} from './createComponent'
export {connect} from './connect'
export {createGate, useGate} from './createGate'
export {useEvent, useList, useStore, useStoreMap, useUnit} from './nossr'
