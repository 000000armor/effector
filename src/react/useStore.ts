import {Store, is, clearNode, createStore} from 'effector'
import React from 'react'
import {useIsomorphicLayoutEffect} from './useIsomorphicLayoutEffect'
import {throwError} from './throw'

export function useStore<State>(store: Store<State>): State {
  if (!is.store(store)) throwError('expect useStore argument to be a store')
  const currentStore = React.useRef(store)
  const setState = React.useState(store.getState())[1]
  useIsomorphicLayoutEffect(() => {
    if (currentStore.current === store) {
      const currentState = store.getState()
      setState(
        typeof currentState === 'function' ? () => currentState : currentState,
      )
    }
    currentStore.current = store
    return store.updates.watch(newState => {
      // check if function in store
      if (typeof newState === 'function') {
        newState = () => newState
      }
      setState(newState)
    })
  }, [store])
  return store.getState()
}

export function useStoreMap<State, Result, Keys extends ReadonlyArray<any>>({
  store,
  keys,
  fn,
}: {
  store: Store<State>
  keys: Keys
  fn(state: State, keys: Keys): Result
}): Result {
  if (!is.store(store)) throwError('useStoreMap expects a store')
  if (!Array.isArray(keys)) throwError('useStoreMap expects an array as keys')
  if (typeof fn !== 'function') throwError('useStoreMap expects a function')
  const result: Store<Result> = React.useMemo(
    () =>
      createStore(fn(store.getState(), keys)).on(store, (_, state) =>
        fn(state, keys),
      ),
    keys,
  )
  const state = useStore(result)
  useIsomorphicLayoutEffect(
    () => () => {
      result.off(store)
      clearNode(result, {deep: true})
    },
    keys,
  )
  return state
}
