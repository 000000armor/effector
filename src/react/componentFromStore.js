//@flow

import * as React from 'react'
import {
  createEvent,
  createStore,
  type Event,
  type Store,
  type Subscription,
} from 'effector'

export function componentFromStore<Props: {...}>(
  propsToVdom: (props: Store<Props>) => Store<React.Node>,
) {
  class ComponentFromStore extends React.Component<Props, {vdom: React.Node, ...}> {
    state /*: {vdom: React.Node, ...}*/ = {vdom: null}

    propsEmitter: Event<Props> = createEvent()

    // Stream of props
    props$: Store<Props> = createStore((null: any))
      .on(this.propsEmitter, (_, props) => props)
      .map(props => {
        if (props) return props
        /*::return props*/
      })

    // Stream of vdom
    vdom$ = propsToVdom(this.props$)

    subscription: Subscription

    componentWillMount() {
      // Subscribe to child prop changes so we know when to re-render
      this.subscription = this.vdom$.watch(vdom => {
        this.setState({vdom})
      })
      this.propsEmitter(this.props)
    }

    componentWillReceiveProps(nextProps: Props) {
      // Receive new props from the owner
      this.propsEmitter(nextProps)
    }

    shouldComponentUpdate(
      nextProps: Props,
      nextState /*: {vdom: React.Node, ...}*/,
    ) {
      return nextState.vdom !== this.state.vdom
    }

    componentWillUnmount() {
      this.props$.off(this.propsEmitter)
      this.subscription.unsubscribe()
    }

    render() {
      return this.state.vdom
    }
  }
  return ComponentFromStore
}
