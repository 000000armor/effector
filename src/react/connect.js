//@flow

import * as React from 'react'

import type {Store} from 'effector'
import {createReactState} from './createReactState'

export function connect<State: Object, Com: React.ComponentType<*>>(
  Component: Com,
): (
  store: Store<State>,
) => React.ComponentType<$Exact<$Diff<React.ElementConfig<Com>, State>>> {
  return store => createReactState(store, Component)
}
