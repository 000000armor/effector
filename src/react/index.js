//@flow

import * as React from 'react'

import type {Store} from 'effector'

export function createReactState<State>(
 store: Store<State>,
): {
 Provider: React.ComponentType<{children: React.Node}>,
 Consumer: React.ComponentType<{
  children: (current: {
   value: State,
   ref: Store<State>,
  }) => React.Node,
 }>,
} {
 const {Provider, Consumer} = React.createContext({
  value: store.getState(),
  ref: store,
 })

 class StoreProvider extends React.Component<
  {
   children: React.Node,
  },
  {
   current: {
    value: State,
    ref: Store<State>,
   },
  },
 > {
  state = {
   current: {
    value: store.getState(),
    ref: store,
   },
  }
  componentDidMount() {
   store.watch(value => this.setState({current: {value, ref: store}}))
  }
  render() {
   return <Provider value={this.state.current}>{this.props.children}</Provider>
  }
 }

 return {
  Provider: StoreProvider,
  Consumer,
 }
}

export function createStoreComponent<
 State: Object,
 Com: React.ComponentType<*>,
>(
 store: Store<State>,
 Component: Com,
): React.ComponentType<$Exact<$Diff<React.ElementConfig<Com>, State>>> {
 const {Provider, Consumer} = createReactState(store)
 const ConnectedComponent = props => (
  <Provider>
   <Consumer>{({value}) => <Component {...props} {...value} />}</Consumer>
  </Provider>
 )
 ConnectedComponent.displayName = `Connect(${Component.displayName
  || 'Unknown'})`
 return ConnectedComponent
}

export function connect<State: Object, Com: React.ComponentType<*>>(
 Component: Com,
): (
 store: Store<State>,
) => React.ComponentType<$Exact<$Diff<React.ElementConfig<Com>, State>>> {
 return store => createStoreComponent(store, Component)
}
