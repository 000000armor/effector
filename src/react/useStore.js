//@flow

import {type Store, is} from 'effector'
import {useReducer, useEffect, useLayoutEffect} from 'react'

const stateReducer = (_: any, payload: any) => payload

const useIsomorphicLayoutEffect =
  typeof window !== 'undefined' ? useLayoutEffect : useEffect

export function useStore<State>(store: Store<State>): State {
  if (!is.store(store)) throw Error('expect useStore argument to be a store')
  const dispatch = useReducer(stateReducer, undefined, store.getState)[1]
  useIsomorphicLayoutEffect(() => store.watch(dispatch), [store])
  return store.getState()
}
