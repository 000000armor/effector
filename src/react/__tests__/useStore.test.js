//@flow

import * as React from 'react'
//$todo
import {render, cleanup, act} from 'react-testing-library'
import {createStore, createEvent} from 'effector'
import {useStore, useStoreFast} from '../useStore'

afterEach(cleanup)

describe('useStore', () => {
  it('should render', () => {
    const store = createStore('foo')
    const changeText = createEvent('change text')
    store.on(changeText, (_, e) => e)

    const Display = props => {
      const state = useStore(store)
      return <span>Store text: {state}</span>
    }

    const {container} = render(<Display />)
    expect(container.firstChild).toMatchSnapshot()
    act(() => {
      changeText('bar')
    })
    // flushEffects()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should throw', () => {
    const ErrorDisplay = props => {
      const state = useStore(undefined)
      return <span>Store text: {state}</span>
    }

    expect(() => {
      render(<ErrorDisplay />)
    }).toThrowErrorMatchingSnapshot()
  })
})

test('useStoreFast', () => {
  const store = createStore('foo')
  const changeText = createEvent('change text')
  store.on(changeText, (_, e) => e)

  const Display = props => {
    const state = useStoreFast(store)
    return <span>Store text: {state}</span>
  }

  const {container} = render(<Display />)
  expect(container.firstChild).toMatchSnapshot()
  act(() => {
    changeText('bar')
  })
  // flushEffects()
  expect(container.firstChild).toMatchSnapshot()
})
