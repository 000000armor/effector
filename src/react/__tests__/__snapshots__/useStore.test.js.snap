// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useStore should render 1`] = `
<span>
  Store text: 
  foo
</span>
`;

exports[`useStore should render 2`] = `
<span>
  Store text: 
  bar
</span>
`;

exports[`useStore should throw 1`] = `"expect useStore argument to be a store"`;
